pkg/cloudlogging/puller.go

```diff
@@ -0,0 +1,156 @@
+	ListLogEntries(context.Context, *loggingpb.ListLogEntriesRequest) *logging.LogEntryIterator
+}
+
+// Puller pulls log entries of GCP organizations, folders, projects, and
+// billingAccounts.
+type Puller struct {
+	client LoggingClient
+	// Required. Name of a parent resource from which to retrieve log entries:
+	//
+	// *  `projects/[PROJECT_ID]`
+	// *  `organizations/[ORGANIZATION_ID]`
+	// *  `billingAccounts/[BILLING_ACCOUNT_ID]`
+	// *  `folders/[FOLDER_ID]`
+	//
+	// May alternatively be one or more views:
+	//
+	//   - `projects/[PROJECT_ID]/locations/[LOCATION_ID]/buckets/[BUCKET_ID]/views/[VIEW_ID]`
+	//   - `organizations/[ORGANIZATION_ID]/locations/[LOCATION_ID]/buckets/[BUCKET_ID]/views/[VIEW_ID]`
+	//   - `billingAccounts/[BILLING_ACCOUNT_ID]/locations/[LOCATION_ID]/buckets/[BUCKET_ID]/views/[VIEW_ID]`
+	//   - `folders/[FOLDER_ID]/locations/[LOCATION_ID]/buckets/[BUCKET_ID]/views/[VIEW_ID]`
+	resource string
+	// Optional retry backoff strategy, default is 5 attempts with fibonacci
+	// backoff that starts at 500ms.
+	retry retry.Backoff
+}
+
+// Option is the option to set up an log puller.
+type Option func(h *Puller) *Puller
+
+// WithRetry provides retry strategy to the log puller.
+func WithRetry(b retry.Backoff) Option {
+	return func(p *Puller) *Puller {
+		p.retry = b
+		return p
+	}
+}
+
+// NewPuller creates a new Puller with provided clients and options.
+func NewPuller(ctx context.Context, c LoggingClient, resource string, opts ...Option) *Puller {
+	p := &Puller{client: c, resource: resource}
+	for _, opt := range opts {
+		p = opt(p)
+	}
+
+	if p.retry == nil {
+		p.retry = retry.WithMaxRetries(5, retry.NewFibonacci(500*time.Millisecond))
+	}
+	return p
+}
+
+// Pull pulls logNum of log entries given log filter.
+func (p *Puller) Pull(ctx context.Context, filter string, logNum int) ([]*loggingpb.LogEntry, error) {
+	var ls []*loggingpb.LogEntry
+	req := &loggingpb.ListLogEntriesRequest{
+		ResourceNames: []string{p.resource},
+		Filter:        filter,
+		// Set descending time order so that newest logs will be returned.
+		OrderBy: "timestamp desc",
+		// Set a large pagesize to optimize read efficiency, see ref
+		// https://cloud.google.com/logging/docs/reference/api-overview#entries-list.
+		PageSize: 1000,
+	}
+	if err := retry.Do(ctx, p.retry, func(ctx context.Context) error {
+		it := p.client.ListLogEntries(ctx, req)
+		for {
+			l, err := it.Next()
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+			if err != nil {
+				return retry.RetryableError(fmt.Errorf("failed to get next log entry: %w, retrying", err))
+			}
+			if len(ls) == logNum {
+				break;
+			}
+			ls = append(ls, l)
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to list log entries: %w", err)
+	}
+	return ls, nil
+}
+
+// SteamPull pulls live log entries, it stops pulling after logNum of log entries.
+func (p *Puller) SteamPull(ctx context.Context, filter string, logNum int) ([]*loggingpb.LogEntry, error) {
```

Comment:
@yolocs to communicate the idea (haven't updated test yet), are you suggesting to have both Pull and StreamPull?